### ISSUE TITLE
missing AZURE_DEFAULTS_SUBSCRIPTION environment var?

### ISSUE THREAD

**Issue by jordan-da** [Azure/azure-cli#17989]
I find it strange the default subscription is missing from the defaults section in:
https://docs.microsoft.com/en-us/cli/azure/azure-cli-configuration#cli-configuration-values-and-environment-variables

Was this done on purpose? It makes the `az` cli hard (and a little unsafe) to use in a multiple subscription environment with standard tools such as [direnv](https://github.com/direnv/direnv).

**Comment by yonzhan** [comment]
@jiasli for awareness

**Comment by patrickjahns** [comment]
Short question on this - does the `AZURE_DEFAULTS_SUBSCRIPTION` variable exist and it was just not documented, or is it not working at all?

**Comment by jiasli** [comment]
`AZURE_DEFAULTS_SUBSCRIPTION` doesn't exist at all.

The information of default subscription is stored in `~/.azure/azureProfile.json`. Look for the subscription with `"isDefault": true` which will be used as the default subscription. 

There are two ways to change subscription

1. `az account set --subscription <sub_name_or_id>`
2. provide `--subscription` argument.

https://github.com/Azure/azure-cli/blob/3f230e856a407043102e99b82d4d22f679111a3c/src/azure-cli-core/azure/cli/core/commands/arm.py#L335-L374
There is no way to set default subscription with environment variable. If this is necessary, we will consider this as a feature request.

**Comment by patrickjahns** [comment]
@jiasli 

Thank you very much for your Feedback 🙏 

I would appreciate having the option to select SUBSCRIPTIONS via environment variables ( `AZURE_DEFAULTS_SUBSCRIPTION` )  

Let me quickly iterate my use-case here - in our setup, we are having identical environments which are separated/segregated on a subscription level. Very often I need to automate tasks and perform the same az cli command on various subscriptions (in parallel).
Being able to open a shell that corresponds to a subscription via performing a `export AZURE_DEFAULTS_SUBSCRIPTION=<sub-first>` and another shell with `export AZURE_DEFAULTS_SUBSCRIPTION=<sub-second>` improve the workflow for me.
Been using that principle for various other cloud providers - and this would fill the gap for me on working azure.

Another use-case would be - to more easily switch between subscriptions - similar to how tools like [`kubectx`](https://github.com/ahmetb/kubectx) or [`kubie`](https://github.com/sbstp/kubie) work for kubectl work

Let me know if you have any questions/suggestions here

**Comment by jordan-da** [comment]
Clearly you don't want everything expressed as environment vars but most other cloud cli's allow you to stipulate high level project scopey stuff as environment vars; zones, regions, projects, etc.. subscription I feel is required, especially since you can specify group.

In what multiple subscription situation would the group environment var `AZURE_DEFAULTS_GROUP` be useful without a `AZURE_DEFAULTS_SUBSCRIPTION` counterpart?

**Comment by jiasli** [comment]
Thanks for the detailed explanation. I understand your use-case. So the main purpose of `AZURE_DEFAULTS_SUBSCRIPTION` is for **parallel execution in multiple shells/terminals**.

On problem with Azure CLI is that it currently doesn't support parallel execution/concurrent build (https://github.com/Azure/azure-cli/issues/9427) with the default config dir `~/.azure`. Even if Azure CLI is run with `AZURE_DEFAULTS_SUBSCRIPTION`, it will crash when `~/accessToken.json` is refreshed in parallel. 

On the other hand, if you run Azure CLI commands serially, `AZURE_DEFAULTS_SUBSCRIPTION` has no advantage over `az account set`.

The best solution for now is

1. Set [`AZURE_CONFIG_DIR`](https://docs.microsoft.com/en-us/cli/azure/use-cli-effectively#concurrent-builds) environment variable to a separate directory for each subscription, like `~/azcli-sub-a`
2. Run `az login` to add login credential to the config dir `~/azcli-sub-a`
3. Run `az account set` to set the default subscription

The same process can be repeated so that you can run `az` commands for separated/segregated subscriptions. 

Once CLI starts to support parallel execution, we will consider adding `AZURE_DEFAULTS_SUBSCRIPTION`.

**Comment by jordan-da** [comment]
It had not occurred to me that you need to re-login when you change subscriptions, in that case I would agree to simply have a config directory per subscription, which pairs nicely with direnv.

> On the other hand, if you run Azure CLI commands serially, AZURE_DEFAULTS_SUBSCRIPTION has no advantage over az account set.

So this is moot now. Thanks!

**Comment by jiasli** [comment]
> It had not occurred to me that you need to re-login when you change subscriptions

Yes. You don't need to re-login when you change subscription. Azure CLI finds all subscriptions you have access to during `az login`, then you can switch subscriptions smoothly with `az account set` - no `az login` required again.

I am glad to know `AZURE_DEFAULTS_SUBSCRIPTION` works for you.

**Comment by patrickjahns** [comment]
@jiasli 

Appreciate your Feedback

> On problem with Azure CLI is that it currently doesn't support parallel execution/concurrent build (#9427) with the default config dir ~/.azure. Even if Azure CLI is run with AZURE_DEFAULTS_SUBSCRIPTION, it will crash when ~/accessToken.json is refreshed in parallel.

This is quite sad to hear/see - I am having various use-cases where I perform things in parallel or retrieve the same information from different environments ( which are basically exact clones of each other) in different subscriptions.

For the operations on AWS, I usually have 3-4 windows open which correlate to an environment. The environments usually have `AWS_PROFILE` set - and this is also visible in the bash prompt - so I know exactly in which environment I am performing actions.
In that workflow I can perform a lookup/operation on Environment 1 - switch shell window and perform the same lookup / operation in Environment 2 - without the need to always switch between `az account set` or have to use `--subscription` as az cli flag.

I am not particularly fond of having a global state that can get modified for any process that is running. 

Instead of having `AZURE_DEFAULTS_SUBSCRIPTION` - would it be possible to have something like `AZURE_SUBSCRIPTION` as environment variable that is similar to `AWS_PROFILE` ?

**Comment by jiasli** [comment]
@patrickjahns, won't [`AZURE_CONFIG_DIR`](https://docs.microsoft.com/en-us/cli/azure/use-cli-effectively#concurrent-builds) environment variable suit your use case? It has more control than `AZURE_SUBSCRIPTION`. It can make Azure CLI run with different identity, subscription, config, etc.

**Comment by mikkelhegn** [comment]
Another use-case for an env. var to indicate the existence of the Azure CLI and an Azure subscription: https://github.com/starship/starship/pull/3189#discussion_r741449006

**Comment by jiasli** [comment]
Once https://github.com/Azure/azure-cli/issues/10241 is implemented, the default subscription can be selected with `AZURE_SUBSCRIPTION_ID`.

**Comment by EpicWink** [comment]
> won't [AZURE_CONFIG_DIR](https://docs.microsoft.com/en-us/cli/azure/use-cli-effectively#concurrent-builds) environment variable suit your use case? It has more control than AZURE_SUBSCRIPTION. It can make Azure CLI run with different identity, subscription, config, etc.

In my use-case, I specifically want to use the same identity and config, but switch subscription. I don't want to keep multiple configs in sync. I simply want an environment-variable alternative for `--subscription`. My current method (in Bash) is:

```shell
az <commands and args> $(if [ "$AZURE_SUBSCRIPTION" ]; then echo "--subscription '$AZURE_SUBSCRIPTION'"; fi)
```

---

I prefer the name `AZURE_SUBSCRIPTION`:
* `AZURE_DEFAULTS_SUBSCRIPTION` suggests it overrides the default: while this may be technically true, I really only want to set `AZURE_SUBSCRIPTION` for a few commands, often switching within the same terminal session.
* `AZURE_SUBSCRIPTION_ID` suggests the subscription name can't be used.

**Comment by jamesikanos** [comment]
Not to be a "Me too" but honestly a AZURE_SUBSCRIPTION_ID env would be so useful. I have long running deployments and logging into a separate config is overkill for just simply scoping a session to a single subscription id.

My script has 70+ az cli calls and maintaining a --subscription flag for each is not feasible and easily missed during maintenance of the scripts.
